### PR TITLE
Add `endBefore` to RTDB queries

### PR DESCRIPTION
**Pull request by IanWyszynski**
[go/rtdb-exclusive-query-filters](go/rtdb-exclusive-query-filters)

### CODE REVIEW

**Comment by IanWyszynski** [comment]
@schmidt-sebastian When you're looking at https://github.com/firebase/firebase-js-sdk/pull/4232 could you also have a look here?

**Comment by IanWyszynski** [comment]
@schmidt-sebastian This is ready for another review round.

File: packages/database/src/api/Query.ts [hunk 1]
```diff
@@ -572,7 +575,28 @@ export class Query {
...
+
+    return new Query(this.repo, this.path, newParams, this.orderByCalled_);
+  }
+
+  endBefore(
+    value: number | string | boolean | null = null,
+    name?: string | null
+  ): Query {
+    validateArgCount('Query.endBefore', 0, 2, arguments.length);
+    validateFirebaseDataArg('Query.endBefore', 1, value, this.path, true);
```
Comment (schmidt-sebastian) [inline packages/database/src/api/Query.ts:591]: This is not optional. Do `null` values still work if you pass `false` as the last argument?

File: packages/database/test/query.test.ts [hunk 2]
```diff
@@ -1667,6 +1891,64 @@ describe('Query Tests', () => {
     );
   });
 
+  it('Ensure startAt / endBefore with priority and name works (2).', () => {
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:1894]: If you need more things on your plate, can you also use a different suffix for this test name?

File: packages/database/test/query.test.ts [hunk 3]
```diff
@@ -3766,6 +4192,90 @@ describe('Query Tests', () => {
...
+    ref.set(
+      {
+        1: true,
+        50: true,
+        550: true,
+        6: true,
+        600: true,
+        70: true,
+        8: true,
+        80: true
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:4232]: You don't need these extra values.

File: packages/database/test/query.test.ts [hunk 4]
```diff
@@ -3766,6 +4192,90 @@ describe('Query Tests', () => {
...
+        1: true,
+        50: true,
+        550: true,
+        6: true,
+        600: true,
+        70: true,
+        8: true,
+        80: true,
+        'a': true
+      },
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:4256]: It might make sense to extract a local variable here so you can use the same set twice.

**Review by schmidt-sebastian — APPROVED** [review]

**Pull request merged**